### PR TITLE
Add ESS resampler to avoid resampling every iteration

### PR DESCRIPTION
**Pull request by jihodori**
Adding ESS rampler and setting the default threshold to 0.5.

Tested with `runtests.jl`

```
Test Summary: | Pass  Total  Time
implemented   |    2      2  0.0s
WARNING: redefinition of constant Main.A. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant Main.B. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant Main.W. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant Main.V. This may fail, cause incorrect answers, or produce other errors.
....................................................................................................Test Summary: |Time
example       | None  0.4s
Test Summary:   | Pass  Total  Time
domain_specific |    7      7  0.2s
Test Summary: | Pass  Total  Time
beliefs       |   24     24  0.1s
Test Summary: | Pass  Total  Time
infer         |    2      2  0.0s
Test Summary:  | Pass  Total  Time
pomdp terminal |    1      1  0.4s
Test Summary: | Pass  Total  Time
alpha         |    2      2  0.4s
```

### CODE REVIEW

File: src/basic.jl [hunk 1]
```diff
@@ -43,20 +43,24 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
 """
 function particle_memory end
 
-function update(up::BasicParticleFilter, b::ParticleCollection, a, o)
+function update(up::BasicParticleFilter, b::ParticleCollection, a, o, τ = 0.5)
```
Comment (zsunberg) [inline src/basic.jl]: The resampling threshhold (tau) should be part of the solver rather than a keyword argument.

File: src/basic.jl [hunk 2]
```diff
@@ -43,20 +43,24 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
...
+    if (calculate_ess(wm) < τ)
+        return resample(up.resampler,
+                        WeightedParticleBelief(pm, wm, sum(wm), nothing),
+                        up.predict_model,
+                        up.reweight_model,
+                        b, a, o,
+                        up.rng)
+    end
+
+    return ParticleCollection(up._particle_memory)
```
Comment (zsunberg) [inline src/basic.jl]: I don't think this second return is correct... doesn't it need to return a `WeightedParticleBelief` if there is no resampling?

**Review by zsunberg — CHANGES_REQUESTED** [review]
Thanks for the PR! I think we need to make some bigger changes for this to work correctly. See #68.

File: src/basic.jl [hunk 3]
```diff
@@ -43,20 +48,25 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
...
+        resampled_particle_collection = resample(
+            up.resampler,
+            WeightedParticleBelief(pm, wm, sum(wm), nothing),
+            up.predict_model,
+            up.reweight_model,
+            b, a, o,
+            up.rng)
+		num_particles = n_particles(resampled_particle_collection)
+        return WeightedParticleBelief(resampled_particle_collection.particles, fill(1.0 / num_particles, num_particles))
+    end
```
Comment (jihodori) [inline src/basic.jl]: actually, if you resample before `reweight!`, the update function does not work as expected

Comment (jihodori) [reply src/basic.jl]: now resampling at the end

File: src/basic.jl (hunk 2 above)
Comment (jihodori) [reply src/basic.jl]: The function now returns `WeightedParticleBelief`

File: src/basic.jl (hunk 1 above)
Comment (jihodori) [reply src/basic.jl]: the threshold is now part of the solver and the default value is set to 0.9 so that it does not deviate too much from the current behavior

File: src/basic.jl (hunk 3 above)
Comment (zsunberg) [reply src/basic.jl]: I don't quite understand this decision. It seems much better to resample before prediction and I don't understand why it didn't work.

**Review by zsunberg — CHANGES_REQUESTED** [review]
I think this looks pretty good, but please address my inline comment about resampling before predicting.

(also, I am seeing many things that I don't like about the code I wrote - it seems like it might be time for a major revamp after this is merged :smile: )

File: src/basic.jl (hunk 3 above)
Comment (jihodori) [reply src/basic.jl]: > I don't quite understand this decision. It seems much better to resample before prediction and I don't understand why it didn't work.

I handled the edge case (first iteration) by checking `wm` is empty. The `resample` function should not be called when the `update` function is called for the first time

File: test/runtests.jl [hunk 4]
```diff
@@ -75,7 +75,7 @@ POMDPs.transition(::TerminalPOMDP, s, a) = Deterministic(s + a)
     pomdp = TerminalPOMDP()
     pf = BootstrapFilter(pomdp, 100)
     bp = update(pf, initialize_belief(pf, Categorical([0.5, 0.5])), -1, 1.0)
-    @test all(particles(bp) .== 1)
+    @test abs(mean(bp) - 1.0) < 1e-5
```
Comment (zsunberg) [inline test/runtests.jl]: Wouldn't it be better to test that all particles with nonzero weight are 1? or `pdf(bp, 1.0)` is approximately 1?

File: src/basic.jl [hunk 5]
```diff
@@ -43,20 +48,25 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
 """
 function particle_memory end
 
-function update(up::BasicParticleFilter, b::ParticleCollection, a, o)
+function update(up::BasicParticleFilter, b::AbstractParticleBelief, a, o)
     pm = up._particle_memory
     wm = up._weight_memory
+    if (!isempty(wm) && calculate_ess(wm) < up.resampling_threshold)
```
Comment (zsunberg) [inline src/basic.jl]: Why do we need to check if `wm` is empty? Shouldn't `calculate_ess` prevent resampling on the first step? Also, the same updater might be re-used for multiple simulations, so this is not a reliable way to check.

(In general, to write maintainable code, it is not good to check implicit side effects (e.g. `wm` will probably be empty at the beginning, but that is not guaranteed) it is better to explicitly check whether it is the first step :smile: but, by design, that is not possible)

File: src/basic.jl [hunk 6]
```diff
@@ -142,3 +152,19 @@ function reweight(m, b, args...)
...
 reweight(f::BasicParticleFilter, args...) = reweight(f.reweight_model, args...)
+
+
+"""
+This function computes Effective Sample Size (ESS) which estimates the number of particles that are effectively contributing to the approximation of the target distribution. 
+
+ESS is divided by `num_particles` to make it easier to get a uniform threshold (scale between 0 and 1) for resampling across different particle filters.
+
+M. S. Arulampalam, S. Maskell, N. Gordon and T. Clapp, "A tutorial on particle filters for online nonlinear/non-Gaussian Bayesian tracking," in IEEE Transactions on Signal Processing, vol. 50, no. 2, pp. 174-188, Feb. 2002, doi: 10.1109/78.978374.
+keywords: {Tutorial;Particle filters;Nonlinear dynamical systems;Costs;Signal processing;Bayesian methods;Particle tracking;Kalman filters;Filtering;Monte Carlo methods},
```
Comment (zsunberg) [inline src/basic.jl]: Delete these keywords?

File: src/basic.jl [hunk 7]
```diff
@@ -142,3 +152,19 @@ function reweight(m, b, args...)
...
+"""
+function calculate_ess(weights)
```
Comment (zsunberg) [inline src/basic.jl]:
```suggestion
function normalized_ess(weights)
```
"calculate" does not really add anything to the name of this function. Also, If someone just said "ESS", does it mean a number between 0 and 1 or an actual number of particles? If ESS means a number of particles, I think we should call this function `normalized_ess`.

**Review by zsunberg — CHANGES_REQUESTED** [review]
This looks really good. I just have one important comment about checking the weights, and a few minor code comments.

File: src/basic.jl (hunk 7 above)
Comment (jihodori) [reply src/basic.jl]: changed the name to `normalized_ess`

File: src/basic.jl (hunk 6 above)
Comment (jihodori) [reply src/basic.jl]: deleted

File: test/runtests.jl (hunk 4 above)
Comment (jihodori) [reply test/runtests.jl]: > Wouldn't it be better to test that all particles with nonzero weight are 1

I agree with this. Addressed this in the newer commit

File: src/basic.jl (hunk 5 above)
Comment (jihodori) [reply src/basic.jl]: Sorry it took me some time to figure out why I had this conditional statement. I realized I was testing this after calling the `resize!` function with an empty list (the initial case), which causes wm to have actual small weights such that `calculate_ess` is always lower than the `resampling_threshold`. 

Now `calculate_ess` is called before the `resize!` function and prevents resampling on the first step as it returns `Inf`
